src/psyclone/psyir/nodes/acc_directives.py

```diff
@@ -801,17 +787,58 @@ def begin_string(self):
         writers_list = sorted(list(writers - readwrites))
         readwrites_list = sorted(list(readwrites))
         if readers_list:
-            str_readers = [str(sig) for sig in readers_list]
-            result += " copyin({0})".format(",".join(str_readers))
+            result += " copyin({0})".format(
+                ",".join(self._create_access_list(readers_list, var_accesses)))
         if writers_list:
-            str_writers = [str(sig) for sig in writers_list]
-            result += " copyout({0})".format(",".join(str_writers))
+            result += " copyout({0})".format(
+                ",".join(self._create_access_list(writers_list, var_accesses)))
         if readwrites_list:
-            str_readwrites = [str(sig) for sig in readwrites_list]
-            result += " copy({0})".format(",".join(str_readwrites))
+            result += " copy({0})".format(",".join(
+                self._create_access_list(readwrites_list, var_accesses)))
 
         return result
 
+    def _create_access_list(self, signatures, var_accesses):
+        '''
+        Constructs a list of variables for inclusion in a data-access clause.
+
+        :param signatures: the list of Signatures for which to create entries \
+            in the list.
+        :type signatures: list of :py:class:`psyclone.core.Signature`
+        :param var_accesses: object holding details on all variable accesses \
+            in the region to which the data-access clause applies.
+        :type var_accesses: :py:class:`psyclone.core.VariablesAccessInfo`
+
+        :returns: list of variable accesses.
+        :rtype: list of str
+
+        '''
+        access_list = []
+        for sig in signatures:
+            if sig.is_structure:
+                # We have to do a 'deep copy' of any structure access. This
+                # means that if we have an access `a%b%c(i)` then we need to
+                # copy `a`, `a%b` and then `a%b%c`.
```

Comment:
Oh, I tried it an apparently is not solved yet :( There is the calls to "acc routine"-annotated subroutines which the manual version does not have that may be the culprit. I will investigate this at some point.

Reply:
Oh, I celebrated prematurely :-( It may be you can force the compiler to inline them and then this problem goes away (there's a flag...perhaps -Minline or something).